// chapa-nodejs/dist/chapa-nodejs.esm.js
import axios from 'axios';
import { customAlphabet } from 'nanoid';
import { alphanumeric } from 'nanoid-dictionary';
import { object, string, array, number, mixed } from 'yup';

function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
      u = i.value;
  } catch (n) {
    return void e(n);
  }
  i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
      e = arguments;
    return new Promise(function (r, o) {
      var a = n.apply(t, e);
      function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }
      function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }
      _next(void 0);
    });
  };
}
function _construct(t, e, r) {
  if (_isNativeReflectConstruct()) return Reflect.construct.apply(null, arguments);
  var o = [null];
  o.push.apply(o, e);
  var p = new (t.bind.apply(t, o))();
  return r && _setPrototypeOf(p, r.prototype), p;
}
function _getPrototypeOf(t) {
  return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) {
    return t.__proto__ || Object.getPrototypeOf(t);
  }, _getPrototypeOf(t);
}
function _inheritsLoose(t, o) {
  t.prototype = Object.create(o.prototype), t.prototype.constructor = t, _setPrototypeOf(t, o);
}
function _isNativeFunction(t) {
  try {
    return -1 !== Function.toString.call(t).indexOf("[native code]");
  } catch (n) {
    return "function" == typeof t;
  }
}
function _isNativeReflectConstruct() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {}));
  } catch (t) {}
  return (_isNativeReflectConstruct = function () {
    return !!t;
  })();
}
function _regeneratorRuntime() {
  _regeneratorRuntime = function () {
    return e;
  };
  var t,
    e = {},
    r = Object.prototype,
    n = r.hasOwnProperty,
    o = Object.defineProperty || function (t, e, r) {
      t[e] = r.value;
    },
    i = "function" == typeof Symbol ? Symbol : {},
    a = i.iterator || "@@iterator",
    c = i.asyncIterator || "@@asyncIterator",
    u = i.toStringTag || "@@toStringTag";
  function define(t, e, r) {
    return Object.defineProperty(t, e, {
      value: r,
      enumerable: !0,
      configurable: !0,
      writable: !0
    }), t[e];
  }
  try {
    define({}, "");
  } catch (t) {
    define = function (t, e, r) {
      return t[e] = r;
    };
  }
  function wrap(t, e, r, n) {
    var i = e && e.prototype instanceof Generator ? e : Generator,
      a = Object.create(i.prototype),
      c = new Context(n || []);
    return o(a, "_invoke", {
      value: makeInvokeMethod(t, r, c)
    }), a;
  }
  function tryCatch(t, e, r) {
    try {
      return {
        type: "normal",
        arg: t.call(e, r)
      };
    } catch (t) {
      return {
        type: "throw",
        arg: t
      };
    }
  }
  e.wrap = wrap;
  var h = "suspendedStart",
    l = "suspendedYield",
    f = "executing",
    s = "completed",
    y = {};
  function Generator() {}
  function GeneratorFunction() {}
  function GeneratorFunctionPrototype() {}
  var p = {};
  define(p, a, function () {
    return this;
  });
  var d = Object.getPrototypeOf,
    v = d && d(d(values([])));
  v && v !== r && n.call(v, a) && (p = v);
  var g = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(p);
  function defineIteratorMethods(t) {
    ["next", "throw", "return"].forEach(function (e) {
      define(t, e, function (t) {
        return this._invoke(e, t);
      });
    });
  }
  function AsyncIterator(t, e) {
    function invoke(r, o, i, a) {
      var c = tryCatch(t[r], t, o);
      if ("throw" !== c.type) {
        var u = c.arg,
          h = u.value;
        return h && "object" == typeof h && n.call(h, "__await") ? e.resolve(h.__await).then(function (t) {
          invoke("next", t, i, a);
        }, function (t) {
          invoke("throw", t, i, a);
        }) : e.resolve(h).then(function (t) {
          u.value = t, i(u);
        }, function (t) {
          return invoke("throw", t, i, a);
        });
      }
      a(c.arg);
    }
    var r;
    o(this, "_invoke", {
      value: function (t, n) {
        function callInvokeWithMethodAndArg() {
          return new e(function (e, r) {
            invoke(t, n, e, r);
          });
        }
        return r = r ? r.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg();
      }
    });
  }
  function makeInvokeMethod(e, r, n) {
    var o = h;
    return function (i, a) {
      if (o === f) throw Error("Generator is already running");
      if (o === s) {
        if ("throw" === i) throw a;
        return {
          value: t,
          done: !0
        };
      }
      for (n.method = i, n.arg = a;;) {
        var c = n.delegate;
        if (c) {
          var u = maybeInvokeDelegate(c, n);
          if (u) {
            if (u === y) continue;
            return u;
          }
        }
        if ("next" === n.method) n.sent = n._sent = n.arg;else if ("throw" === n.method) {
          if (o === h) throw o = s, n.arg;
          n.dispatchException(n.arg);
        } else "return" === n.method && n.abrupt("return", n.arg);
        o = f;
        var p = tryCatch(e, r, n);
        if ("normal" === p.type) {
          if (o = n.done ? s : l, p.arg === y) continue;
          return {
            value: p.arg,
            done: n.done
          };
        }
        "throw" === p.type && (o = s, n.method = "throw", n.arg = p.arg);
      }
    };
  }
  function maybeInvokeDelegate(e, r) {
    var n = r.method,
      o = e.iterator[n];
    if (o === t) return r.delegate = null, "throw" === n && e.iterator.return && (r.method = "return", r.arg = t, maybeInvokeDelegate(e, r), "throw" === r.method) || "return" !== n && (r.method = "throw", r.arg = new TypeError("The iterator does not provide a '" + n + "' method")), y;
    var i = tryCatch(o, e.iterator, r.arg);
    if ("throw" === i.type) return r.method = "throw", r.arg = i.arg, r.delegate = null, y;
    var a = i.arg;
    return a ? a.done ? (r[e.resultName] = a.value, r.next = e.nextLoc, "return" !== r.method && (r.method = "next", r.arg = t), r.delegate = null, y) : a : (r.method = "throw", r.arg = new TypeError("iterator result is not an object"), r.delegate = null, y);
  }
  function pushTryEntry(t) {
    var e = {
      tryLoc: t[0]
    };
    1 in t && (e.catchLoc = t[1]), 2 in t && (e.finallyLoc = t[2], e.afterLoc = t[3]), this.tryEntries.push(e);
  }
  function resetTryEntry(t) {
    var e = t.completion || {};
    e.type = "normal", delete e.arg, t.completion = e;
  }
  function Context(t) {
    this.tryEntries = [{
      tryLoc: "root"
    }], t.forEach(pushTryEntry, this), this.reset(!0);
  }
  function values(e) {
    if (e || "" === e) {
      var r = e[a];
      if (r) return r.call(e);
      if ("function" == typeof e.next) return e;
      if (!isNaN(e.length)) {
        var o = -1,
          i = function next() {
            for (; ++o < e.length;) if (n.call(e, o)) return next.value = e[o], next.done = !1, next;
            return next.value = t, next.done = !0, next;
          };
        return i.next = i;
      }
    }
    throw new TypeError(typeof e + " is not iterable");
  }
  return GeneratorFunction.prototype = GeneratorFunctionPrototype, o(g, "constructor", {
    value: GeneratorFunctionPrototype,
    configurable: !0
  }), o(GeneratorFunctionPrototype, "constructor", {
    value: GeneratorFunction,
    configurable: !0
  }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, u, "GeneratorFunction"), e.isGeneratorFunction = function (t) {
    var e = "function" == typeof t && t.constructor;
    return !!e && (e === GeneratorFunction || "GeneratorFunction" === (e.displayName || e.name));
  }, e.mark = function (t) {
    return Object.setPrototypeOf ? Object.setPrototypeOf(t, GeneratorFunctionPrototype) : (t.__proto__ = GeneratorFunctionPrototype, define(t, u, "GeneratorFunction")), t.prototype = Object.create(g), t;
  }, e.awrap = function (t) {
    return {
      __await: t
    };
  }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, c, function () {
    return this;
  }), e.AsyncIterator = AsyncIterator, e.async = function (t, r, n, o, i) {
    void 0 === i && (i = Promise);
    var a = new AsyncIterator(wrap(t, r, n, o), i);
    return e.isGeneratorFunction(r) ? a : a.next().then(function (t) {
      return t.done ? t.value : a.next();
    });
  }, defineIteratorMethods(g), define(g, u, "Generator"), define(g, a, function () {
    return this;
  }), define(g, "toString", function () {
    return "[object Generator]";
  }), e.keys = function (t) {
    var e = Object(t),
      r = [];
    for (var n in e) r.push(n);
    return r.reverse(), function next() {
      for (; r.length;) {
        var t = r.pop();
        if (t in e) return next.value = t, next.done = !1, next;
      }
      return next.done = !0, next;
    };
  }, e.values = values, Context.prototype = {
    constructor: Context,
    reset: function (e) {
      if (this.prev = 0, this.next = 0, this.sent = this._sent = t, this.done = !1, this.delegate = null, this.method = "next", this.arg = t, this.tryEntries.forEach(resetTryEntry), !e) for (var r in this) "t" === r.charAt(0) && n.call(this, r) && !isNaN(+r.slice(1)) && (this[r] = t);
    },
    stop: function () {
      this.done = !0;
      var t = this.tryEntries[0].completion;
      if ("throw" === t.type) throw t.arg;
      return this.rval;
    },
    dispatchException: function (e) {
      if (this.done) throw e;
      var r = this;
      function handle(n, o) {
        return a.type = "throw", a.arg = e, r.next = n, o && (r.method = "next", r.arg = t), !!o;
      }
      for (var o = this.tryEntries.length - 1; o >= 0; --o) {
        var i = this.tryEntries[o],
          a = i.completion;
        if ("root" === i.tryLoc) return handle("end");
        if (i.tryLoc <= this.prev) {
          var c = n.call(i, "catchLoc"),
            u = n.call(i, "finallyLoc");
          if (c && u) {
            if (this.prev < i.catchLoc) return handle(i.catchLoc, !0);
            if (this.prev < i.finallyLoc) return handle(i.finallyLoc);
          } else if (c) {
            if (this.prev < i.catchLoc) return handle(i.catchLoc, !0);
          } else {
            if (!u) throw Error("try statement without catch or finally");
            if (this.prev < i.finallyLoc) return handle(i.finallyLoc);
          }
        }
      }
    },
    abrupt: function (t, e) {
      for (var r = this.tryEntries.length - 1; r >= 0; --r) {
        var o = this.tryEntries[r];
        if (o.tryLoc <= this.prev && n.call(o, "finallyLoc") && this.prev < o.finallyLoc) {
          var i = o;
          break;
        }
      }
      i && ("break" === t || "continue" === t) && i.tryLoc <= e && e <= i.finallyLoc && (i = null);
      var a = i ? i.completion : {};
      return a.type = t, a.arg = e, i ? (this.method = "next", this.next = i.finallyLoc, y) : this.complete(a);
    },
    complete: function (t, e) {
      if ("throw" === t.type) throw t.arg;
      return "break" === t.type || "continue" === t.type ? this.next = t.arg : "return" === t.type ? (this.rval = this.arg = t.arg, this.method = "return", this.next = "end") : "normal" === t.type && e && (this.next = e), y;
    },
    finish: function (t) {
      for (var e = this.tryEntries.length - 1; e >= 0; --e) {
        var r = this.tryEntries[e];
        if (r.finallyLoc === t) return this.complete(r.completion, r.afterLoc), resetTryEntry(r), y;
      }
    },
    catch: function (t) {
      for (var e = this.tryEntries.length - 1; e >= 0; --e) {
        var r = this.tryEntries[e];
        if (r.tryLoc === t) {
          var n = r.completion;
          if ("throw" === n.type) {
            var o = n.arg;
            resetTryEntry(r);
          }
          return o;
        }
      }
      throw Error("illegal catch attempt");
    },
    delegateYield: function (e, r, n) {
      return this.delegate = {
        iterator: values(e),
        resultName: r,
        nextLoc: n
      }, "next" === this.method && (this.arg = t), y;
    }
  }, e;
}
function _setPrototypeOf(t, e) {
  return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) {
    return t.__proto__ = e, t;
  }, _setPrototypeOf(t, e);
}
function _wrapNativeSuper(t) {
  var r = "function" == typeof Map ? new Map() : void 0;
  return _wrapNativeSuper = function (t) {
    if (null === t || !_isNativeFunction(t)) return t;
    if ("function" != typeof t) throw new TypeError("Super expression must either be null or a function");
    if (void 0 !== r) {
      if (r.has(t)) return r.get(t);
      r.set(t, Wrapper);
    }
    function Wrapper() {
      return _construct(t, arguments, _getPrototypeOf(this).constructor);
    }
    return Wrapper.prototype = Object.create(t.prototype, {
      constructor: {
        value: Wrapper,
        enumerable: !1,
        writable: !0,
        configurable: !0
      }
    }), _setPrototypeOf(Wrapper, t);
  }, _wrapNativeSuper(t);
}

var ChapaUrls;
(function (ChapaUrls) {
  ChapaUrls["INITIALIZE"] = "https://api.chapa.co/v1/transaction/initialize";
  ChapaUrls["MOBILE_INITIALIZE"] = "https://api.chapa.co/v1/transaction/mobile-initialize";
  ChapaUrls["VERIFY"] = "https://api.chapa.co/v1/transaction/verify";
  ChapaUrls["BANK"] = "https://api.chapa.co/v1/banks";
  ChapaUrls["SUBACCOUNT"] = "https://api.chapa.co/v1/subaccount";
  ChapaUrls["TRANSACTION"] = "https://api.chapa.co/v1/transactions";
  ChapaUrls["TRANSACTION_LOG"] = "https://api.chapa.co/v1/transaction/events";
  ChapaUrls["TRANSFER"] = "https://api.chapa.co/v1/transfers";
  ChapaUrls["BULK_TRANSFER"] = "https://api.chapa.co/v1/bulk-transfers";
  ChapaUrls["VERIFY_TRANSFER"] = "https://api.chapa.co/v1/transfers/verify";
  ChapaUrls["DIRECT_CHARGE"] = "https://api.chapa.co/v1/charges";
  ChapaUrls["AUTHORIZE_DIRECT_CHARGE"] = "https://api.chapa.co/v1/validate";
})(ChapaUrls || (ChapaUrls = {}));

var SplitType;
(function (SplitType) {
  SplitType["PERCENTAGE"] = "percentage";
  SplitType["FLAT"] = "flat";
})(SplitType || (SplitType = {}));

var HttpException = /*#__PURE__*/function (_Error) {
  function HttpException(message, status) {
    var _this;
    _this = _Error.call(this, message) || this;
    _this.status = status;
    Error.captureStackTrace(_this, _this.constructor);
    return _this;
  }
  _inheritsLoose(HttpException, _Error);
  return HttpException;
}( /*#__PURE__*/_wrapNativeSuper(Error));

var validateInitializeOptions = /*#__PURE__*/function () {
  var _ref = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          schema = object().shape({
            first_name: string().nullable().optional(),
            last_name: string().nullable().optional(),
            email: string().email().nullable().optional(),
            phone_number: string().matches(/^0[79]\d{8}$/, 'Phone number must be 10 digits and start with 09 or 07').nullable().optional(),
            currency: string().required(),
            amount: string().required(),
            tx_ref: string().required(),
            callback_url: string().nullable().optional(),
            return_url: string().nullable().optional(),
            customization: object().shape({
              title: string().nullable().optional(),
              description: string().nullable().optional(),
              logo: string().nullable().optional()
            }).optional(),
            subaccounts: array().of(object().shape({
              id: string().required(),
              split_type: string().nullable().optional(),
              split_value: string().nullable().optional()
            })).nullable().optional()
          });
          _context.next = 3;
          return schema.validate(options);
        case 3:
          return _context.abrupt("return", _context.sent);
        case 4:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function validateInitializeOptions(_x) {
    return _ref.apply(this, arguments);
  };
}();
var validateVerifyOptions = /*#__PURE__*/function () {
  var _ref2 = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          schema = object().shape({
            tx_ref: string().required()
          });
          _context2.next = 3;
          return schema.validate(options);
        case 3:
          return _context2.abrupt("return", _context2.sent);
        case 4:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function validateVerifyOptions(_x2) {
    return _ref2.apply(this, arguments);
  };
}();

var validateCreateSubaccountOptions = /*#__PURE__*/function () {
  var _ref = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(createSubaccountOptions) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          schema = object().shape({
            business_name: string().required(),
            account_name: string().required(),
            bank_code: number().required(),
            account_number: string().required(),
            split_type: mixed().oneOf(Object.values(SplitType)).required(),
            split_value: number().required()
          });
          _context.next = 3;
          return schema.validate(createSubaccountOptions);
        case 3:
          return _context.abrupt("return", _context.sent);
        case 4:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function validateCreateSubaccountOptions(_x) {
    return _ref.apply(this, arguments);
  };
}();

var validateGetTransactionLogsOptions = /*#__PURE__*/function () {
  var _ref = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          schema = object().shape({
            ref_id: string().required()
          });
          _context.next = 3;
          return schema.validate(options);
        case 3:
          return _context.abrupt("return", _context.sent);
        case 4:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function validateGetTransactionLogsOptions(_x) {
    return _ref.apply(this, arguments);
  };
}();

var validateTransferOptions = /*#__PURE__*/function () {
  var _ref = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          schema = object().shape({
            account_name: string().required(),
            account_number: string().required(),
            amount: string().required(),
            currency: string().required(),
            reference: string().required(),
            bank_code: number().required()
          });
          _context.next = 3;
          return schema.validate(options);
        case 3:
          return _context.abrupt("return", _context.sent);
        case 4:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function validateTransferOptions(_x) {
    return _ref.apply(this, arguments);
  };
}();
var validateBulkTransferOptions = /*#__PURE__*/function () {
  var _ref2 = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          schema = object().shape({
            title: string().required(),
            currency: string().required(),
            bulk_data: array().of(object().shape({
              account_name: string().required(),
              account_number: string().required(),
              amount: string().required(),
              reference: string().required(),
              bank_code: number().required()
            })).required()
          });
          _context2.next = 3;
          return schema.validate(options);
        case 3:
          return _context2.abrupt("return", _context2.sent);
        case 4:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function validateBulkTransferOptions(_x2) {
    return _ref2.apply(this, arguments);
  };
}();

var validateDirectChargeOptions = /*#__PURE__*/function () {
  var _ref = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          schema = object().shape({
            first_name: string().nullable().optional(),
            last_name: string().nullable().optional(),
            email: string().email().nullable().optional(),
            mobile: string().matches(/^0[79]\d{8}$/, 'Phone number must be 10 digits and start with 09 or 07').required(),
            currency: string().required(),
            amount: string().required(),
            tx_ref: string().required(),
            type: string().required()
          });
          _context.next = 3;
          return schema.validate(options);
        case 3:
          return _context.abrupt("return", _context.sent);
        case 4:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function validateDirectChargeOptions(_x) {
    return _ref.apply(this, arguments);
  };
}();
var validateAuthorizeDirectChargeOptions = /*#__PURE__*/function () {
  var _ref2 = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(options) {
    var schema;
    return _regeneratorRuntime().wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          schema = object().shape({
            reference: string().required(),
            client: string().required(),
            type: string().required()
          });
          _context2.next = 3;
          return schema.validate(options);
        case 3:
          return _context2.abrupt("return", _context2.sent);
        case 4:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function validateAuthorizeDirectChargeOptions(_x2) {
    return _ref2.apply(this, arguments);
  };
}();

var Chapa = /*#__PURE__*/function () {
  function Chapa(chapaOptions) {
    this.chapaOptions = chapaOptions;
  }
  var _proto = Chapa.prototype;
  _proto.initialize = /*#__PURE__*/function () {
    var _initialize = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            _context.next = 3;
            return validateInitializeOptions(options);
          case 3:
            _context.next = 5;
            return axios.post(ChapaUrls.INITIALIZE, options, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context.sent;
            return _context.abrupt("return", response.data);
          case 9:
            _context.prev = 9;
            _context.t0 = _context["catch"](0);
            if (!_context.t0.response) {
              _context.next = 15;
              break;
            }
            throw new HttpException(_context.t0.response.data.message, _context.t0.response.status);
          case 15:
            if (!(_context.t0.name === 'ValidationError')) {
              _context.next = 19;
              break;
            }
            throw new HttpException(_context.t0.errors[0], 400);
          case 19:
            throw _context.t0;
          case 20:
          case "end":
            return _context.stop();
        }
      }, _callee, this, [[0, 9]]);
    }));
    function initialize(_x) {
      return _initialize.apply(this, arguments);
    }
    return initialize;
  }();
  _proto.mobileInitialize = /*#__PURE__*/function () {
    var _mobileInitialize = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            _context2.prev = 0;
            _context2.next = 3;
            return validateInitializeOptions(options);
          case 3:
            _context2.next = 5;
            return axios.post(ChapaUrls.MOBILE_INITIALIZE, options, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context2.sent;
            return _context2.abrupt("return", response.data);
          case 9:
            _context2.prev = 9;
            _context2.t0 = _context2["catch"](0);
            if (!_context2.t0.response) {
              _context2.next = 15;
              break;
            }
            throw new HttpException(_context2.t0.response.data.message, _context2.t0.response.status);
          case 15:
            if (!(_context2.t0.name === 'ValidationError')) {
              _context2.next = 19;
              break;
            }
            throw new HttpException(_context2.t0.errors[0], 400);
          case 19:
            throw _context2.t0;
          case 20:
          case "end":
            return _context2.stop();
        }
      }, _callee2, this, [[0, 9]]);
    }));
    function mobileInitialize(_x2) {
      return _mobileInitialize.apply(this, arguments);
    }
    return mobileInitialize;
  }();
  _proto.verify = /*#__PURE__*/function () {
    var _verify = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.prev = 0;
            _context3.next = 3;
            return validateVerifyOptions(options);
          case 3:
            _context3.next = 5;
            return axios.get(ChapaUrls.VERIFY + "/" + options.tx_ref, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context3.sent;
            return _context3.abrupt("return", response.data);
          case 9:
            _context3.prev = 9;
            _context3.t0 = _context3["catch"](0);
            if (!_context3.t0.response) {
              _context3.next = 15;
              break;
            }
            throw new HttpException(_context3.t0.response.data.message, _context3.t0.response.status);
          case 15:
            if (!(_context3.t0.name === 'ValidationError')) {
              _context3.next = 19;
              break;
            }
            throw new HttpException(_context3.t0.errors[0], 400);
          case 19:
            throw _context3.t0;
          case 20:
          case "end":
            return _context3.stop();
        }
      }, _callee3, this, [[0, 9]]);
    }));
    function verify(_x3) {
      return _verify.apply(this, arguments);
    }
    return verify;
  }();
  _proto.genTxRef = /*#__PURE__*/function () {
    var _genTxRef = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee4(options) {
      var _ref, _ref$removePrefix, removePrefix, _ref$prefix, prefix, _ref$size, size, nanoid, reference;
      return _regeneratorRuntime().wrap(function _callee4$(_context4) {
        while (1) switch (_context4.prev = _context4.next) {
          case 0:
            _ref = options || {}, _ref$removePrefix = _ref.removePrefix, removePrefix = _ref$removePrefix === void 0 ? false : _ref$removePrefix, _ref$prefix = _ref.prefix, prefix = _ref$prefix === void 0 ? 'TX' : _ref$prefix, _ref$size = _ref.size, size = _ref$size === void 0 ? 15 : _ref$size;
            nanoid = customAlphabet(alphanumeric, size);
            reference = nanoid();
            if (!removePrefix) {
              _context4.next = 5;
              break;
            }
            return _context4.abrupt("return", reference.toUpperCase());
          case 5:
            return _context4.abrupt("return", prefix + "-" + reference.toUpperCase());
          case 6:
          case "end":
            return _context4.stop();
        }
      }, _callee4);
    }));
    function genTxRef(_x4) {
      return _genTxRef.apply(this, arguments);
    }
    return genTxRef;
  }();
  _proto.getBanks = /*#__PURE__*/function () {
    var _getBanks = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee5() {
      var banks;
      return _regeneratorRuntime().wrap(function _callee5$(_context5) {
        while (1) switch (_context5.prev = _context5.next) {
          case 0:
            _context5.prev = 0;
            _context5.next = 3;
            return axios.get(ChapaUrls.BANK, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 3:
            banks = _context5.sent;
            return _context5.abrupt("return", banks.data);
          case 7:
            _context5.prev = 7;
            _context5.t0 = _context5["catch"](0);
            if (!_context5.t0.response) {
              _context5.next = 13;
              break;
            }
            throw new HttpException(_context5.t0.response.data.message, _context5.t0.response.status);
          case 13:
            throw _context5.t0;
          case 14:
          case "end":
            return _context5.stop();
        }
      }, _callee5, this, [[0, 7]]);
    }));
    function getBanks() {
      return _getBanks.apply(this, arguments);
    }
    return getBanks;
  }();
  _proto.createSubaccount = /*#__PURE__*/function () {
    var _createSubaccount = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee6(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee6$(_context6) {
        while (1) switch (_context6.prev = _context6.next) {
          case 0:
            _context6.prev = 0;
            _context6.next = 3;
            return validateCreateSubaccountOptions(options);
          case 3:
            _context6.next = 5;
            return axios.post(ChapaUrls.SUBACCOUNT, options, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context6.sent;
            return _context6.abrupt("return", response.data);
          case 9:
            _context6.prev = 9;
            _context6.t0 = _context6["catch"](0);
            if (!_context6.t0.response) {
              _context6.next = 15;
              break;
            }
            throw new HttpException(_context6.t0.response.data.message, _context6.t0.response.status);
          case 15:
            if (!(_context6.t0.name === 'ValidationError')) {
              _context6.next = 19;
              break;
            }
            throw new HttpException(_context6.t0.errors[0], 400);
          case 19:
            throw _context6.t0;
          case 20:
          case "end":
            return _context6.stop();
        }
      }, _callee6, this, [[0, 9]]);
    }));
    function createSubaccount(_x5) {
      return _createSubaccount.apply(this, arguments);
    }
    return createSubaccount;
  }();
  _proto.getTransactions = /*#__PURE__*/function () {
    var _getTransactions = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee7() {
      var response;
      return _regeneratorRuntime().wrap(function _callee7$(_context7) {
        while (1) switch (_context7.prev = _context7.next) {
          case 0:
            _context7.prev = 0;
            _context7.next = 3;
            return axios.get(ChapaUrls.TRANSACTION, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 3:
            response = _context7.sent;
            return _context7.abrupt("return", response.data);
          case 7:
            _context7.prev = 7;
            _context7.t0 = _context7["catch"](0);
            if (!_context7.t0.response) {
              _context7.next = 13;
              break;
            }
            throw new HttpException(_context7.t0.response.data.message, _context7.t0.response.status);
          case 13:
            throw _context7.t0;
          case 14:
          case "end":
            return _context7.stop();
        }
      }, _callee7, this, [[0, 7]]);
    }));
    function getTransactions() {
      return _getTransactions.apply(this, arguments);
    }
    return getTransactions;
  }();
  _proto.getTransactionLogs = /*#__PURE__*/function () {
    var _getTransactionLogs = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee8(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee8$(_context8) {
        while (1) switch (_context8.prev = _context8.next) {
          case 0:
            _context8.prev = 0;
            _context8.next = 3;
            return validateGetTransactionLogsOptions(options);
          case 3:
            _context8.next = 5;
            return axios.get(ChapaUrls.TRANSACTION_LOG + "/" + options.ref_id, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context8.sent;
            return _context8.abrupt("return", response.data);
          case 9:
            _context8.prev = 9;
            _context8.t0 = _context8["catch"](0);
            if (!_context8.t0.response) {
              _context8.next = 15;
              break;
            }
            throw new HttpException(_context8.t0.response.data.message, _context8.t0.response.status);
          case 15:
            if (!(_context8.t0.name === 'ValidationError')) {
              _context8.next = 19;
              break;
            }
            throw new HttpException(_context8.t0.errors[0], 400);
          case 19:
            throw _context8.t0;
          case 20:
          case "end":
            return _context8.stop();
        }
      }, _callee8, this, [[0, 9]]);
    }));
    function getTransactionLogs(_x6) {
      return _getTransactionLogs.apply(this, arguments);
    }
    return getTransactionLogs;
  }();
  _proto.transfer = /*#__PURE__*/function () {
    var _transfer = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee9(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee9$(_context9) {
        while (1) switch (_context9.prev = _context9.next) {
          case 0:
            _context9.prev = 0;
            _context9.next = 3;
            return validateTransferOptions(options);
          case 3:
            _context9.next = 5;
            return axios.post(ChapaUrls.TRANSFER, options, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context9.sent;
            return _context9.abrupt("return", response.data);
          case 9:
            _context9.prev = 9;
            _context9.t0 = _context9["catch"](0);
            if (!_context9.t0.response) {
              _context9.next = 15;
              break;
            }
            throw new HttpException(_context9.t0.response.data.message, _context9.t0.response.status);
          case 15:
            if (!(_context9.t0.name === 'ValidationError')) {
              _context9.next = 19;
              break;
            }
            throw new HttpException(_context9.t0.errors[0], 400);
          case 19:
            throw _context9.t0;
          case 20:
          case "end":
            return _context9.stop();
        }
      }, _callee9, this, [[0, 9]]);
    }));
    function transfer(_x7) {
      return _transfer.apply(this, arguments);
    }
    return transfer;
  }();
  _proto.bulkTransfer = /*#__PURE__*/function () {
    var _bulkTransfer = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee10(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee10$(_context10) {
        while (1) switch (_context10.prev = _context10.next) {
          case 0:
            _context10.prev = 0;
            _context10.next = 3;
            return validateBulkTransferOptions(options);
          case 3:
            _context10.next = 5;
            return axios.post(ChapaUrls.BULK_TRANSFER, options, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context10.sent;
            return _context10.abrupt("return", response.data);
          case 9:
            _context10.prev = 9;
            _context10.t0 = _context10["catch"](0);
            if (!_context10.t0.response) {
              _context10.next = 15;
              break;
            }
            throw new HttpException(_context10.t0.response.data.message, _context10.t0.response.status);
          case 15:
            if (!(_context10.t0.name === 'ValidationError')) {
              _context10.next = 19;
              break;
            }
            throw new HttpException(_context10.t0.errors[0], 400);
          case 19:
            throw _context10.t0;
          case 20:
          case "end":
            return _context10.stop();
        }
      }, _callee10, this, [[0, 9]]);
    }));
    function bulkTransfer(_x8) {
      return _bulkTransfer.apply(this, arguments);
    }
    return bulkTransfer;
  }();
  _proto.verifyTransfer = /*#__PURE__*/function () {
    var _verifyTransfer = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee11(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee11$(_context11) {
        while (1) switch (_context11.prev = _context11.next) {
          case 0:
            _context11.prev = 0;
            _context11.next = 3;
            return validateVerifyOptions(options);
          case 3:
            _context11.next = 5;
            return axios.get(ChapaUrls.VERIFY_TRANSFER + "/" + options.tx_ref, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context11.sent;
            return _context11.abrupt("return", response.data);
          case 9:
            _context11.prev = 9;
            _context11.t0 = _context11["catch"](0);
            if (!_context11.t0.response) {
              _context11.next = 15;
              break;
            }
            throw new HttpException(_context11.t0.response.data.message, _context11.t0.response.status);
          case 15:
            if (!(_context11.t0.name === 'ValidationError')) {
              _context11.next = 19;
              break;
            }
            throw new HttpException(_context11.t0.errors[0], 400);
          case 19:
            throw _context11.t0;
          case 20:
          case "end":
            return _context11.stop();
        }
      }, _callee11, this, [[0, 9]]);
    }));
    function verifyTransfer(_x9) {
      return _verifyTransfer.apply(this, arguments);
    }
    return verifyTransfer;
  }();
  _proto.getTransfers = /*#__PURE__*/function () {
    var _getTransfers = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee12() {
      var response;
      return _regeneratorRuntime().wrap(function _callee12$(_context12) {
        while (1) switch (_context12.prev = _context12.next) {
          case 0:
            _context12.prev = 0;
            _context12.next = 3;
            return axios.get(ChapaUrls.TRANSFER, {
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 3:
            response = _context12.sent;
            return _context12.abrupt("return", response.data);
          case 7:
            _context12.prev = 7;
            _context12.t0 = _context12["catch"](0);
            if (!_context12.t0.response) {
              _context12.next = 13;
              break;
            }
            throw new HttpException(_context12.t0.response.data.message, _context12.t0.response.status);
          case 13:
            throw _context12.t0;
          case 14:
          case "end":
            return _context12.stop();
        }
      }, _callee12, this, [[0, 7]]);
    }));
    function getTransfers() {
      return _getTransfers.apply(this, arguments);
    }
    return getTransfers;
  }();
  _proto.directCharge = /*#__PURE__*/function () {
    var _directCharge = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee13(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee13$(_context13) {
        while (1) switch (_context13.prev = _context13.next) {
          case 0:
            _context13.prev = 0;
            _context13.next = 3;
            return validateDirectChargeOptions(options);
          case 3:
            _context13.next = 5;
            return axios.post(ChapaUrls.DIRECT_CHARGE, options, {
              params: {
                type: options.type
              },
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context13.sent;
            return _context13.abrupt("return", response.data);
          case 9:
            _context13.prev = 9;
            _context13.t0 = _context13["catch"](0);
            if (!_context13.t0.response) {
              _context13.next = 15;
              break;
            }
            throw new HttpException(_context13.t0.response.data.message, _context13.t0.response.status);
          case 15:
            if (!(_context13.t0.name === 'ValidationError')) {
              _context13.next = 19;
              break;
            }
            throw new HttpException(_context13.t0.errors[0], 400);
          case 19:
            throw _context13.t0;
          case 20:
          case "end":
            return _context13.stop();
        }
      }, _callee13, this, [[0, 9]]);
    }));
    function directCharge(_x10) {
      return _directCharge.apply(this, arguments);
    }
    return directCharge;
  }();
  _proto.authorizeDirectCharge = /*#__PURE__*/function () {
    var _authorizeDirectCharge = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee14(options) {
      var response;
      return _regeneratorRuntime().wrap(function _callee14$(_context14) {
        while (1) switch (_context14.prev = _context14.next) {
          case 0:
            _context14.prev = 0;
            _context14.next = 3;
            return validateAuthorizeDirectChargeOptions(options);
          case 3:
            _context14.next = 5;
            return axios.post(ChapaUrls.AUTHORIZE_DIRECT_CHARGE, options, {
              params: {
                type: options.type
              },
              headers: {
                Authorization: "Bearer " + this.chapaOptions.secretKey
              }
            });
          case 5:
            response = _context14.sent;
            return _context14.abrupt("return", response.data);
          case 9:
            _context14.prev = 9;
            _context14.t0 = _context14["catch"](0);
            if (!_context14.t0.response) {
              _context14.next = 15;
              break;
            }
            throw new HttpException(_context14.t0.response.data.message, _context14.t0.response.status);
          case 15:
            if (!(_context14.t0.name === 'ValidationError')) {
              _context14.next = 19;
              break;
            }
            throw new HttpException(_context14.t0.errors[0], 400);
          case 19:
            throw _context14.t0;
          case 20:
          case "end":
            return _context14.stop();
        }
      }, _callee14, this, [[0, 9]]);
    }));
    function authorizeDirectCharge(_x11) {
      return _authorizeDirectCharge.apply(this, arguments);
    }
    return authorizeDirectCharge;
  }();
  return Chapa;
}();

export { Chapa, ChapaUrls, SplitType };
